### ISSUE TITLE
[TS] Add TypeScript unit testing pipeline

### ISSUE THREAD

**Issue by alexmirrington** [grommet/grommet#5111]
## Motivation

In the last few weeks, I've noticed a lot of issues outlining discrepancies between actual component bahaviour (outlined in `.js` files) and the types defined in `index.d.ts` files. I propose either migrating existing tests to `.ts` files, or adding TypeScript-specific tests alongside existing tests to help improve the Grommet TypeScript experience for users.

I'd like this issue to be a thread for initial discussion/proposals, and if/when we decide on a way forward I'll log a separate issue to track migration/implementation progress.

## Related Issues

Here is a list ot the types of TypeScript issues that are being raised recently:

- Missing/incorrect typings:
  - #5108
  - #5107 
  - #5117 
- Generics for prop types
  - #5087 
  - #5041
- Storybook TS
  - #5121

**Comment by oskari** [comment]
There really is much to desire for the TypeScript experience. Even new features have issues when using with TS. Grommet really should have first-class TypeScript support.

**Comment by ShimiSun** [comment]
@alexmirrington THANK YOU! As we are not field experts on Typescript, we are always welcome to feedback and suggestions from the community of TS specialists. 
Grommet v2.17.0 that was released recently, had massive TS improvements thanks to the community pushing for it, and we need the community's support to help us improve in the future as well.
Feel free to file any POC PRs that demonstrate how we can improve the Typescript dev experience, so the discussion could be more proactive and we can improve the overall TS experience within Grommet.

**Comment by alexmirrington** [comment]
@ShimiSun What are your thoughts on an achievable direction for improving the grommet-ts experience? There are a few directions to head, some of which are more difficult/time consuming than others:

1. First class TypeScript support, converting all existing `*.js` files to `*.ts`
  - This is by far the best option in terms of pure TS experience, but time consuming to implement and will mak eit difficult for people less experienced with TypeScript to contribute
2. Convert unit tests to TypeScript, and run them with `ts-jest` so that when type checks fail, so do unit tests.  
  - I personally think this is the most achievable option to start out, since many of the current issues we are seeing with TS are things that could have very easily been picked up by failing tests. I'm happy to start work on drafting a sample PR if you think this is a good direction to head.
3. Add TypeScript stories in Storybook to make it more transparent to grommet users/contributors how we see TypeScript being used with Grommet.
  - At the moment, everyone seems to have different ideas on the best way to use types with Grommet since there's no real documentation surrounding it.

Keen to hear your thoughts on this!

**Comment by ShimiSun** [comment]
Thank you @alexmirrington, Great discussion!!

1. We had discussions on #1 in the past, taking this approach will drive a major release with breaking changes, which is not something we would like to commit at this point in terms of efforts and resources.
2. I like this idea and I'm willing to give it a try! Feel free to file a POC PR to start the conversation in depth.
3. We currently have the infrastructure to do it so this might be the easiest way to go, if you think the stories could be written better I would love to see a POC of what you have in mind.

**Comment by ericsoderberghp** [comment]
I also like option #2, perhaps focused more on API surface than on detailed interaction sequences. This seems like it would be easy to incrementally add to, start small and grow as needed.

**Comment by alexmirrington** [comment]
The above PR adds support for unit tests in `.tsx` files, as per my second point. If you're happy with these changes, the way forward is to create a separate tracking issue that will allow us to refactor all `.js` unit tests to `.tsx` files. This will be tedious, but essential for improved TS support. We will also come across a lot of bugs regarding prop types, etc. that will need to be addressed as we convert unit tests to TS to ensure builds don't fail.

**Comment by alexmirrington** [comment]
@ShimiSun Any updates on the PR? Would love to get a second pair of eyes on the PR so we can start improving Grommet's TS support through test migrations :)

**Comment by alexmirrington** [comment]
@ShimiSun @ericsoderberghp @taysea Any updates on this? I know it has been a while but this is something I would still like to see in Grommet. If you would prefer I closed this issue and created another one, I'd be more than happy to.

**Comment by jcfilben** [comment]
Hi @alexmirrington sorry for the long wait on this. Thanks for your patience and following up on the issue. I think typescript testing would be a beneficial addition to grommet. @ericsoderberghp do you know if there was a reason we were holding off on this or has it just been waiting for someone to take a stab at reviewing it?

**Comment by alexmirrington** [comment]
@jcfilben I've just created a new PR and closed the old one, as the dependencies for the old PR were quite a few months out of date. All technical details for reviewers can be found in the description of the new PR.
